### ISSUE TITLE
Add some useful Go utilities for an upcoming analyzer driver [phab:D406]

### ISSUE THREAD

**Issue by kythe** [google/kythe#636]
_Created by **schroederc** at 2015-07-15 18:03:34_:

**Comment by kythe** [comment]
_**schroederc** wrote at 2015-07-15 18:04:58_:

Fatal -> Error

**Comment by kythe** [comment]
_**zarko** wrote at 2015-07-15 18:11:45_:

//kythe/go/util/process:process_test fails on OS X:

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
--- FAIL: TestAsyncSuccess (5.01s)
	process_test.go:32: Process was nil
	process_test.go:52: Process exited with unknown error: fork/exec /bin/true: no such file or directory (<nil>)
	process_test.go:38: Process exited with unknown error: fork/exec /bin/true: no such file or directory (<nil>)
	process_test.go:62: Process did not finish before timeout
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x8 pc=0x77ac5]

goroutine 9 [running]:
os.(*ProcessState).success(0x0, 0x20828ef58)
	/usr/local/go/src/os/exec_posix.go:73 +0x5
os.(*ProcessState).Success(0x0, 0x48952)
	/usr/local/go/src/os/doc.go:72 +0x28
kythe.io/kythe/go/util/process.func·008(0x0, 0x2208284bb0, 0x2082ac660)
	/private/var/tmp/_bazel_zarko/2be6994bf77f75bac8f1433eb6e9dc33/kythe-ci/kythe/go/util/process/process_test.go:87 +0x3d
kythe.io/kythe/go/util/process.(*Async).runHandlers(0x2082ac5a0, 0x2208284bb0, 0x2082ac660)
	/private/var/tmp/_bazel_zarko/2be6994bf77f75bac8f1433eb6e9dc33/kythe-ci/kythe/go/util/process/async.go:70 +0xbf
kythe.io/kythe/go/util/process.func·009()
	/private/var/tmp/_bazel_zarko/2be6994bf77f75bac8f1433eb6e9dc33/kythe-ci/kythe/go/util/process/async.go:55 +0x40
created by kythe.io/kythe/go/util/process.(*Async).Start
	/private/var/tmp/_bazel_zarko/2be6994bf77f75bac8f1433eb6e9dc33/kythe-ci/kythe/go/util/process/async.go:59 +0x13a

goroutine 1 [chan receive]:
testing.RunTests(0x1ad290, 0x229920, 0x3, 0x3, 0x2082aa000)
	/usr/local/go/src/testing/testing.go:556 +0xad6
testing.(*M).Run(0x2082a80a0, 0x232480)
	/usr/local/go/src/testing/testing.go:485 +0x6c
testing.Main(0x1ad290, 0x229920, 0x3, 0x3, 0x232480, 0x0, 0x0, 0x232480, 0x0, 0x0)
	/usr/local/go/src/testing/testing.go:454 +0x192
main.main()
	/private/var/tmp/_bazel_zarko/2be6994bf77f75bac8f1433eb6e9dc33/kythe-ci/tools/go/testmain.go:23 +0x96

goroutine 8 [select]:
kythe.io/kythe/go/util/process.TestAsyncError(0x2082d4090)
	/private/var/tmp/_bazel_zarko/2be6994bf77f75bac8f1433eb6e9dc33/kythe-ci/kythe/go/util/process/process_test.go:100 +0x547
testing.tRunner(0x2082d4090, 0x229938)
	/usr/local/go/src/testing/testing.go:447 +0xbf
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:555 +0xa8b
```

**Comment by kythe** [comment]
_**zarko** required changes at 2015-07-15 18:12:03_

**Comment by kythe** [comment]
_**schroederc** wrote at 2015-07-15 18:29:07_:

Lookup true/false binary paths

**Comment by kythe** [comment]
_**schroederc** wrote at 2015-07-15 18:29:18_:

Better?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-07-15 19:46:58_:

Doesn't this lead to a race with any other processes waiting to be assigned an ephemeral port? We went through a whole Thing in //kythe/cxx/common/testdata/start_http_service.sh to deal with this via a temporary file. The documentation indicates that the port is "likely unused", but the test later on asserts that the port must be unused.

**Comment by kythe** [comment]
_**zarko** wrote at 2015-07-15 19:46:58_:

The test now passes.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-07-15 19:50:27_:

It is true that this will be racy, but I don't know of a better way to handle this problem.  Most of the time this will work.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-07-15 20:15:24_:

OK; maybe it would be good to change the name of the function then ("PickProbablyUnusedPort"?) and change the test to retry the pick/listen process some number of times with random backoff before giving up.

Ideally we'd be able to appeal to some third party to broker ports for us (or do some other well-founded thing) but doing that properly and in a cross-platform way seems like it's out of scope for this diff.

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2015-07-15 20:15:24_

**Comment by kythe** [comment]
_**schroederc** wrote at 2015-07-16 15:48:10_:

Remove flaky parts of test

**Comment by kythe** [comment]
_**schroederc** closed via commit 673a5dafcd5d7d746fdc39003c7c5a5ef812aadf at 2015-07-16 15:48:29_
